warmup/summary_statistics.py

```diff
@@ -0,0 +1,299 @@
+    # different machines.
+    assert len(data_dictionaries) == 1
+    machine = data_dictionaries.keys()[0]
+    summary_data = { machine: dict() }
+    # Parse data dictionaries.
+    keys = sorted(data_dictionaries[machine]['wallclock_times'].keys())
+    for key in sorted(keys):
+        wallclock_times = data_dictionaries[machine]['wallclock_times'][key]
+        if len(wallclock_times) == 0:
+            print ('WARNING: Skipping: %s from %s (no executions)' %
+                   (key, machine))
+        elif len(wallclock_times[0]) == 0:
+            print('WARNING: Skipping: %s from %s (benchmark crashed)' %
+                  (key, machine))
+        else:
+            bench, vm, variant = key.split(':')
+            if vm not in summary_data[machine].keys():
+                summary_data[machine][vm] = list()
+            # Get information for all p_execs of this key.
+            categories = list()
+            steady_state_means = list()
+            steady_iters = list()
+            time_to_steadys = list()
+            n_pexecs = len(data_dictionaries[machine]['wallclock_times'][key])
+            # Lists of changepoints, outliers and segment means for each process execution.
+            changepoints, outliers, segments = list(), list(), list()
+            for p_exec in xrange(n_pexecs):
+                changepoints.append(data_dictionaries[machine]['changepoints'][key][p_exec])
+                segments.append(data_dictionaries[machine]['changepoint_means'][key][p_exec])
+                outliers.append(data_dictionaries[machine]['all_outliers'][key][p_exec])
+                categories.append(data_dictionaries[machine]['classifications'][key][p_exec])
+                # Next we calculate the iteration at which a steady state was
+                # reached, it's average segment mean and the time to reach a
+                # steady state. However, the last segment may be equivalent to
+                # its adjacent segments, so we first need to know which segments
+                # are steady-state segments.
+                first_steady_segment = len(data_dictionaries[machine]['changepoint_means'][key][p_exec]) - 1
+                num_steady_segments = 1
+                last_segment_mean = data_dictionaries[machine]['changepoint_means'][key][p_exec][-1]
+                lower_bound = min(last_segment_mean * (1.0 - half_bound), last_segment_mean - delta)
+                upper_bound = max(last_segment_mean * (1.0 + half_bound), last_segment_mean + delta)
+                for index in xrange(len(data_dictionaries[machine]['changepoint_means'][key][p_exec]) - 2, -1, -1):
+                    current_segment_mean = data_dictionaries[machine]['changepoint_means'][key][p_exec][index]
+                    if current_segment_mean >= lower_bound and current_segment_mean <= upper_bound:
+                        first_steady_segment -= 1
+                        num_steady_segments += 1
```

Comment:
That may be so, but I think the two variables are slightly clearer, and this code is a little tricky already.